### ISSUE TITLE
Memory Optimizations for Spatial Analysis

### ISSUE THREAD

**Issue by ackagel** [angelolab/ark-analysis#337]
## **Description**
Suppose that someone were to have a FOV with ~20,000 cells.  The distance matrix generated for this FOV will take up a lot of memory.  In particular, for float64 (worst case), this distance matrix will take up 20,000^2 * 64 / 8 = **3.2 GB** of memory.  This isn't terrible on its own, but `compute_close_cell_num_random` is also fairly memory intensive, scaling linearly in memory with bootstrap number.  This leads to some users encountering 'out of memory errors' and dying kernels.

## **Proposed Solutions**
There a few optimizations (some easier than others) that can reduce the memory requirements:
 - The distance matrix is computed using `cdist` from scipy, which outputs float64.  We can shave off half of the occupied memory by changing this to float32.
 - Output array, `close_num_rand` doesn't need to be int32 (uint16 is probs sufficient)
 - Needless creation of xarray in compute_close_cell_num_rand
 - Both the distance matrix and `close_num_rand` are symmetric.  Numpy doesn't have a memory efficient data structure for exploiting symmetry.  Depending on the efficacy of other fixes, a quick class/object could be written for symmetric data structures, with numpy under the hood.

## **Alternate Solution**
Dynamic distance matrix computation.  Instead of storing the distance matrix in memory, it can be computed only when it's needed.  `cdist` is very fast, so recomputing the full distance matrix isn't a big deal.  This does create some code gymnastics on the backend, but user experience could improve.

**Comment by alex-l-kong** [comment]
@ackagel does `close_num` need to be `int32`? Pretty sure we can shave off even more memory by making `close_num` `uint16` as well.

**Comment by ngreenwald** [comment]
No, I think it would be extremely rare for an FOV to have more than 5k cells. Although we'll likely be having more data that's 2048x2048, which would further increase the cell count. I think first step is changing the default dtypes, and seeing how far that gets us. 

Given that almost everyone is going to be able to run the pipeline on a single FOV, I think the next logical step is smarter batching of multiple FOVs. If someone wants to analyze their cohort of data, it's fine to serially analyze one FOV at a time. So we can need to rewrite how the higher level functions iterate through each FOV to reduce the amount in memory at any one time.

**Comment by ackagel** [comment]
I only bring it up because Candace tried running a FOV with 20,000ish cells and encountered this slowdown.  She said that she had a FOV with 80,000 cells too 😬

**Comment by alex-l-kong** [comment]
@ngreenwald I'll open a separate issue for smarter batching of multiple FOVs.

**Comment by ngreenwald** [comment]
I used this to debug some stuff in DeepCell: https://pypi.org/project/memory-profiler/. I recommend before we start changing any code, we figure out which steps are actually causing the largest memory load

**Comment by ngreenwald** [comment]
Yes, either way it makes sense to reduce the dtypes. We'll likely need to add a step when we load the deepcell_output that checks what the max label val is. If it's above the max val of int16, we'll use int32.

**Comment by ackagel** [comment]
I'll go ahead and reduce the dtypes in a PR later today.  That can close this issue, and I'll post another issue adding memory-profiler so we can do general optimizations.

**Comment by ngreenwald** [comment]
Candace's 80k FOV is a large tiled image of multiple FOVs. I don't think there's any particular reason to think running a distance matrix on an entire tiled image is necessary for now
